### PR TITLE
Full Code Review

### PR DESCRIPTION
**Pull request by ianobermiller**

### CODE REVIEW

File: src/cats/CatSlideshow.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+      doFetch(pageIndex);
+    }
+
+    if (prefetchPageIndex !== pageIndex && !pages[prefetchPageIndex]) {
+      doFetch(prefetchPageIndex);
+    }
+  }, [index, pages, selectedBreedID, fetchPage]);
+
+  useEffect(() => {
+    const prefetchMap = prefetchMapRef.current;
```
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/CatSlideshow.js:164]: Once of the nice things that hooks allow is pulling out this logic into a custom hook, which would be a good exercise and would make this component more readable.

```tsx
const {clearPrefetchMap} = usePrefetchImages(pages, index);
```

File: src/cats/CatSlideshow.js [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        if (url) {
+          prefetchMap.set(i, "prefetch");
+
+          preloadImage(url)
+            .then(() => updateImageStatus(i, "ready"))
+            .catch(() => updateImageStatus(i, "error"));
+        }
+      });
+    }
+  }, [pages, index, visibleIndex, state.indexReadyMap]);
```
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/CatSlideshow.js:204]: visibleIndex isn't used either

File: src/cats/CatSlideshow.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      Math.max(0, state.index + increment),
+      state.maxIndex
+    );
+    const isReady = state.indexReadyMap[newIndex] === "ready";
+
+    return {
+      ...state,
+      direction: increment > 0 ? Direction.Next : Direction.Previous,
+      index: newIndex,
+      visibleIndex: isReady ? newIndex : state.visibleIndex,
```
Comment (ianobermiller) [inline src/cats/CatSlideshow.js:90]: Ah this is neat, so you keep showing the previous one until the new one loads

File: src/cats/SlideAnimation.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+const Direction = {
+  Next: 0,
+  Previous: 1,
+};
+
+function SlideAnimation({ child, direction }) {
+  const [slideChildren, setSlideChildren] = useState([]);
+
+  useEffect(() => {
+    setSlideChildren((prevChildren) => {
```
Comment (ianobermiller) [inline src/cats/SlideAnimation.js:13]: this is a bit hard to follow, but I've wrote a similar component recently, `Crossfade` and this logic is somewhat complicated anyway

**Review by ianobermiller — COMMENTED** [review]
Overall I thought this was really good! I think the major thing I'd try to refactor is to avoid using `useEffect` to react to changes in data (some places, like `SlideAnimation` are ok) and instead change state directly in the events.

File: src/cats/BreedSelector.css [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+.BreedSelector form label {
+  color: #999;
+  margin-right: 20px;
+  position: relative;
+  top: 1px;
+}
+
+.BreedSelector .selectContainer {
+  position: relative;
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.css]: vestigal, thanks!

File: src/cats/BreedSelector.css [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+.BreedSelector form select {
+  border: 2px solid #ccc;
+  font-size: 1.2rem;
+  width: 200px;
+}
+
+.BreedSelector form select.isError {
+  background-color: #ccc;
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.css]: good idea. added to my todo list

File: src/cats/BreedSelector.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import "./BreedSelector.css";
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.js]: done

File: src/cats/BreedSelector.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+function BreedSelector({ dispatch, breeds, selectedBreedID, status }) {
+  function onSelectChange(e) {
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.js]: I didn't realize that subtlety about built-in components. Thx

File: src/cats/BreedSelector.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    <div className="BreedSelector">
+      <form>
+        <label>Breed</label>
+        <span className="selectContainer">
+          <select
+            className={status === "error" ? "isError" : ""}
+            disabled={breeds === null || status === "error"}
+            onChange={onSelectChange}
+            type="select"
+            value={selectedBreedID || undefined}
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.js:31]: The ID is a string (e.g. "all"). It is initialized to null in the parent component, but the select wants it to be a valid value or undefined. I don't like initializing to undefined, so this is the workaround to avoid the error.

File: src/cats/Cards.css [hunk 10]
```diff
@@ -0,0 +1,23 @@
+.MessageCard,
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/Cards.css]: done. One question I have about CSS modules. If you look at the updated code, MessageCard is a specialization of BaseCard. I can compose the CSS for both of them together either by passing an additional className from MessageCard to BaseCard, or I can uses composes:  in the CSS module to basically accomplish the same thing. 

It's not entirely appropriate in this use case, but how do you think about one of these approaches vs the other?

File: src/cats/Cards.css [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+.LoadingCard {
+  height: 300px;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+}
+
+.LoadingCard {
+  background-color: rgba(255, 255, 255, 0.4);
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/Cards.css]: cool thanks

File: src/cats/CatApp.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+import { MessageCard, LoadingCard } from "./Cards.js";
+import "./CatApp.css";
+
+const BREED_ID_KEY = "CatApp-breed-id";
+
+function catAppReducer(state, action) {
+  //console.log("%ccatapp: " + action.type, "color: green", action);
+  switch (action.type) {
+    case "select-breed": {
+      window.localStorage.setItem(BREED_ID_KEY, action.id);
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatApp.js]: I wired it up to useLocalStateStorage which existing in the codebase but was not being used.

File: src/cats/CatApp.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+        const json = await response.json();
+        let breeds = [{ id: "all", name: "All Breeds" }].concat(
+          json.filter((b) => ({ id: b.id, name: b.name }))
+        );
+
+        dispatch({
+          type: "breeds-loaded",
+          breeds,
+          selectedBreedID: window.localStorage.getItem(BREED_ID_KEY) || "all",
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatApp.js]: good idea. Done

File: src/cats/CatSlideshow.js [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+  return pages?.[pageIndex]?.data?.[offset]?.url;
+}
+
+function initCatSlideshowState() {
+  return {
+    direction: Direction.Next,
+    index: 0,
+    maxIndex: null,
+    visibleIndex: null,
+    readyMap: {},
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js]: good catch!

File: src/cats/CatSlideshow.js (hunk 3 above)
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js:97]: yes. I like how this turned out : )

File: src/cats/CatSlideshow.js [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+    }
+    default:
+      throw new Error(`Unhandled action type: ${action.type}`);
+  }
+}
+
+function CatSlideshow({ selectedBreedID }) {
+  const prevSelectedBreedIDRef = useRef(selectedBreedID);
+  const prefetchMapRef = useRef(new Map());
+  const { pages, metadata, fetchPage, resetPages } = usePaginatedFetch(
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js:134]: Sure. I will make those changes and see how the code works out. I think I learned from my studies that reducers have some really solid properties (stable dispatch function, clarity, testability) and so I tended to use them all over the place. 

What's your decision process on when to use a reducer?
1) If you need to have logic that updates multiple pieces of dependent state
2) when you want to avoid passing several callback props to a child
3) If you end up with duplicated state update logic in your component
4) ...?

File: src/cats/CatSlideshow.js [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+  const { direction, index, visibleIndex } = state;
+
+  useEffect(() => {
+    if (selectedBreedID !== prevSelectedBreedIDRef.current) {
+      prevSelectedBreedIDRef.current = selectedBreedID;
+      resetPages(selectedBreedID);
+      dispatch({ type: "change-breed", selectedBreedID });
+      prefetchMapRef.current.clear();
+    }
+  }, [selectedBreedID, resetPages]);
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js]: I guess it's probably [this](https://overreacted.io/a-complete-guide-to-useeffect/)

File: src/cats/CatSlideshow.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js:171]: Cool, I will do that.

File: src/cats/CatSlideshow.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js]: I guess the lint plugin is great at finding missing deps but not so great at finding unused deps.

File: src/cats/getCatsApiFetchParams.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+//let API_KEY = "1a2691b0-d86c-41c1-b968-3b264d96ff7";
+let API_KEY = "DEMO-API-KEY";
+const API_VERSION = "1";
+const SERVER = "api.thecatapi.com";
+
+const getCatsApiFetchParams = (path, queryParams = {}, slowMillis) => {
+  const params = new URLSearchParams(queryParams);
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/getCatsApiFetchParams.js:11]: what specifically do you like here?

File: src/cats/useLocalStateStorage.js [hunk 18]
```diff
@@ -0,0 +1,37 @@
+import { useEffect, useState } from "react";
+
+function useLocalStateStorage(key, initialValue) {
+  const [state, setState] = useState(() => {
+    return initialValue || JSON.parse(window.localStorage.getItem(key));
+  });
+
+  useEffect(() => {
+    window.localStorage.setItem(key, JSON.stringify(state));
+  }, [state, key]);
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/useLocalStateStorage.js]: yup

File: src/cats/usePaginatedFetch.js [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+  const resetPages = useCallback(
+    (key) => {
+      abortFetch();
+      dispatch({ type: "reset", key });
+    },
+    [abortFetch]
+  );
+
+  const fetchPage = useCallback(
+    (url, options, index, key, getPageData, getMetadata = null) => {
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/usePaginatedFetch.js]: done

File: src/cats/CatSlideshow.js [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+  const [state, dispatch] = useReducer(
+    catSlideshowReducer,
+    undefined,
+    initCatSlideshowState
+  );
+
+  const { direction, index, visibleIndex } = state;
+
+  useEffect(() => {
+    if (selectedBreedID !== prevSelectedBreedIDRef.current) {
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js:148]: Since selectedBreedID is coming through via props, I don't need a useEffect at all. I can just delete line 119 and 126. Do I have that right?

File: src/cats/CatApp.js [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+      />
+      {loadStatus === "loading" || state.selectedBreedID === null ? (
+        <>
+          <LoadingCard />
+          <CatSlideshowControls isDisabled={true} />
+        </>
+      ) : loadStatus === "error" ? (
+        <>
+          <MessageCard className="error">an error has occurred</MessageCard>
+          <CatSlideshowControls isDisabled={true} />
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatApp.js:67]: I agree with you guys that CatSlideshow should own rendering the controls.

@chadwhiting, rather than pass loadStatus into the Cards, I'd prefer to keep them dumb and continue to have CatApp manage their display.

File: src/cats/CatSlideshow.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js:134]: So I pulled out the two reducers, as you suggest. It made a huge difference in clarity to usePaginatedFetch and it's clearly better. It also reminded me that the functional version of setState is super useful inside async responses so you don't get bit by stale state.

It did also improve CatSlideshow but the difference is less marked. For components that need to reset all their state occasionally, one function can be used to initialize the reducer and also reset the state (see line 48). After replacing the reducer with 5 useState calls, the reset function feels somewhat duplicative. Do you have a suggestion on how to improve that?

File: src/cats/CatSlideshowControls.js [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+        className={styles.button}
+        disabled={isDisabled || !canScrollLeft}
+        onClick={() => dispatch({ type: "decrement-index" })}
+      >
+        &lt;
+      </button>
+      <button
+        className={styles.button}
+        disabled={isDisabled || !canScrollRight}
+        onClick={() => dispatch({ type: "increment-index" })}
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshowControls.js:21]: I ended up replacing this code so that a function wrapped is useCallback comes from CatSlideshow and is passed unchanged to the button elements. I think that accomplishes the same thing.

File: src/cats/CatSlideshow.js [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+import getCatsApiFetchParams from "./getCatsApiFetchParams.js";
+import usePaginatedFetch from "./usePaginatedFetch";
+import CatSlideshowControls from "./CatSlideshowControls.js";
+import { SlideAnimation, Direction } from "./SlideAnimation.js";
+import { LoadingCard } from "./Cards.js";
+import styles from "./CatSlideshow.module.css";
+
+const PAGE_SIZE = 20;
+const PAGE_PREFETCH = 8;
+const MAX_PREFETCH = 4;
```
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.js]: I guess there's no need to use context for this because these values never change during the app lifetime.

File: src/cats/CatSlideshowControls.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/CatSlideshowControls.js]: Worth noting there is no benefit to memoizing callbacks passed to raw elements -- React does the same amount of work either way

File: src/cats/CatSlideshow.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/CatSlideshow.js]: I'll take another look at how you are doing it now, but in general if you have to check for when props are changing, it is possible to instead react to the signal that caused the prop to change in the first place. (e.g. instead of having a prop named `scrollTop` in which you look for changes, you could call an imperative method in the scroll event itself)

File: src/cats/CatSlideshow.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/CatSlideshow.js]: Yes, `useReducer` is definitely good and you found good resources on that. I used it recently to drive the entire state of a virtualized list and it worked out really well and was highly testable.

File: src/cats/CatSlideshow.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/CatSlideshow.js]: Not quite, he actually recommended using `useRef` to keep track of the previous value explicitly and then bail in the `useEffect` if it hasn't changed.

File: src/cats/getCatsApiFetchParams.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/getCatsApiFetchParams.js]: Just the use of `URLSearchParams`, the browser gives you so much these days!

File: src/cats/useAbortableFetch.js [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+function useAbortableFetch(initialStatus) {
+  const state = useAsync(initialStatus);
+  const { status, data, error, run: asyncRun, abort: asyncAbort } = state;
+  const controllerRef = useRef(new AbortController());
+
+  const abort = useCallback(() => {
+    controllerRef.current.abort();
+    controllerRef.current = new AbortController();
+    asyncAbort();
+  }, [asyncAbort, controllerRef]);
```
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/useAbortableFetch.js]: I only meant if you use an empty array the lint rule will add back in `asyncAbort`, and you should keep it even if you know it is stable. The perf hit is minimal and its better to follow the conservative lint rule.

File: src/cats/CatSlideshow.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/CatSlideshow.js]: Yeah, that is true. Sometimes it is because they can be legitimate (you want the effect to rerun when a dependency is changed, even if it isn't used directly)

File: src/cats/Cards.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (ianobermiller) [reply src/cats/Cards.css]: I definitely don't like passing in the additional classname. I would say that they both have css modules, and MessageCard would pass an extra class to BaseCard to apply its styles.

File: src/cats/BreedSelector.tsx [hunk 25]
```diff
@@ -0,0 +1,66 @@
+import { useCallback, useMemo } from "react";
+import clsx from "clsx";
+import { Status } from "./useAsync";
+import styles from "./BreedSelector.module.css";
+
+type BreedType = {
```
Comment (ianobermiller) [inline src/cats/BreedSelector.tsx:6]: nit: don't sufix types with `Type`. This can sometimes cause conflicts, and typically I rename the component instead of the type, since the type is used more.

File: src/cats/BreedSelector.tsx [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+  id: string;
+  name: string;
+};
+
+type BreedSelectorPropsType = {
```
Comment (ianobermiller) [inline src/cats/BreedSelector.tsx:11]: nit: always name the props for the main component `Props`. It is a good signal and the main thing people are looking for when they open a component they might want to use.

Less important, but we recently switched to using `interface` everywhere as there is a performance benefit (`interface` is opaque to the typechecker so it does less work, whereas `type` is always expanded). That being said I prefer the `type` syntax from using Flow for so long, so don't put much stock in this.

File: src/cats/BreedSelector.tsx [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+  onSelectBreedID: (id: string) => void;
```
Comment (ianobermiller) [inline src/cats/BreedSelector.tsx:12]: nit: there is a more concise alternative, but it is just stylistic:

```tsx
onSelectBreed(id: string): void;
```

File: src/cats/BreedSelector.tsx [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+}: BreedSelectorPropsType) {
+  const isError = status === "error";
+  const onSelectChange = useCallback(
+    (e: React.ChangeEvent<HTMLSelectElement>) => {
+      onSelectBreedID(e.target.value);
+    },
+    [onSelectBreedID]
+  );
+
+  const selectOptions = useMemo(() => {
```
Comment (ianobermiller) [inline src/cats/BreedSelector.tsx:32]: just like with memoizing callbacks, you don't need to memoize the children passed to a primitive like `select`

File: src/cats/Cards.tsx [hunk 29]
```diff
@@ -0,0 +1,63 @@
+import { ReactNode } from "react";
+import clsx from "clsx";
+import styles from "./Cards.module.css";
+
+type BaseCardProps = {
+  className?: string;
+  children: ReactNode;
+};
+
+type LoadingCardProps = {
```
Comment (ianobermiller) [inline src/cats/Cards.tsx:10]: nit: move these next to the component

File: src/cats/Cards.tsx [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+
+function ErrorCard({ className, children }: ErrorCardProps) {
+  return (
+    <MessageCard className={className} messageClassName={styles.errorMessage}>
+      {children}
+    </MessageCard>
+  );
+}
+
+export { LoadingCard, MessageCard, ErrorCard };
```
Comment (ianobermiller) [inline src/cats/Cards.tsx:63]: nit: personal preference, but I prefer to export these where they are defined

File: src/cats/CatApp.tsx [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+import BreedSelector from "./BreedSelector";
+import { ErrorCard, LoadingCard } from "./Cards";
+import styles from "./CatApp.module.css";
+
+type CatAppPropsType = {
+  className?: string;
+};
+
+type BreedsAPIResponseType = Array<{ id: string; name: string }>;
+type BreedsType = BreedsAPIResponseType | null;
```
Comment (ianobermiller) [inline src/cats/CatApp.tsx:16]: nit: IMO pulling this out hinders readability, I'd just inline this type

File: src/cats/CatApp.tsx [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+          }
+          setBreeds(breeds);
+        }
+      })
+      .catch((error) => {
+        // suppress error
+      });
+  }, [config.selectedBreedID, runFetch, setAppConfig]);
+
+  const onSelectBreedID = useCallback<(breedID: string) => void>(
```
Comment (ianobermiller) [inline src/cats/CatApp.tsx:50]: nit: put the type on the function inside and let TypeScript infer it:

```tsx
const onSelectBreedID = useCallback(
  (breedID: string) => {
  ...
);

File: src/cats/CatApp.tsx [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+    (breedID) => {
+      setAppConfig({ ...config, selectedBreedID: breedID });
+    },
+    [config, setAppConfig]
+  );
+
+  return (
+    <div className={clsx([styles.catApp, className])}>
```
Comment (ianobermiller) [inline src/cats/CatApp.tsx:58]: nice, this looks much cleaner

File: src/cats/CatApp.tsx [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+    const { url, options } = getCatsApiFetchParams("/breeds");
+    runFetch(url, options)
+      .then(async (response: Response | void) => {
+        if (response !== undefined) {
+          const json: BreedsAPIResponseType = await response.json();
+          let breeds = [{ id: "all", name: "All Breeds" }].concat(
+            json.map((b) => ({ id: b.id, name: b.name }))
+          );
+
+          if (breeds.find((b) => b.id === config.selectedBreedID)) {
```
Comment (ianobermiller) [inline src/cats/CatApp.tsx:39]: I'm a bit confused -- should this be negated?

File: src/cats/CatSlideshow.tsx [hunk 35]
```diff
@@ -0,0 +1,178 @@
...
+    SlideshowPageType,
+    string
+  >(selectedBreedID, "loading");
+  const { imageLoadMap, resetPrefetch } = usePrefetchImages(
+    index,
+    useCallback((i) => getImageURL(pages, i), [pages])
+  );
+
+  useEffect(() => {
+    if (selectedBreedID !== previousBreedID) {
```
Comment (ianobermiller) [inline src/cats/CatSlideshow.tsx:86]: this was definitely better with useReducer

File: src/cats/CatSlideshow.tsx [hunk 36]
```diff
@@ -0,0 +1,178 @@
...
+
+    if (!pages[pageIndex]) {
+      doFetch(pageIndex);
+    }
+
+    if (prefetchPageIndex !== pageIndex && !pages[prefetchPageIndex]) {
+      doFetch(prefetchPageIndex);
+    }
+
+    if (metadata?.paginationCount !== undefined) {
```
Comment (ianobermiller) [inline src/cats/CatSlideshow.tsx:110]: this should be in its own effect -- or ideally in the reducer itself

File: src/cats/CatSlideshow.tsx [hunk 37]
```diff
@@ -0,0 +1,178 @@
...
+
+export default function CatSlideshow({
+  selectedBreedID,
+}: CatSlideshowPropsType) {
+  const [direction, setDirection] = useState(Direction.Next);
+  const [index, setIndex] = useState<number>(0);
+  const [visibleIndex, setVisibleIndex] = useState<number | null>(null);
+  const [maxIndex, setMaxIndex] = useState<number | null>(null);
+  const previousBreedID = usePrevious<string>(selectedBreedID);
+  const { pages, metadata, status, fetchPage, resetPages } = usePaginatedFetch<
```
Comment (ianobermiller) [inline src/cats/CatSlideshow.tsx:76]: So going back to the earlier comment, I think what I would do here is to restore the reducer in `CatSlideshow` and have an option for `usePaginatedFetch` to call callbacks. Then you can just wire the up directly to a dispatch on your CatSlideshow reducer instead of trying to update CatSlideshow state in response to pagination state changing

File: src/cats/SlideAnimation.tsx [hunk 38]
```diff
@@ -0,0 +1,59 @@
+import { useEffect, useState } from "react";
+import clsx from "clsx";
+import styles from "./SlideAnimation.module.css";
+
+enum Direction {
```
Comment (ianobermiller) [inline src/cats/SlideAnimation.tsx:5]: this is also an opinion, but I find that a simple string union is usually nicer than enums:

```
type Direction = 'previous' | 'next';
```

TypeScript is really good about the typing inside of switch, === comparison, and of course assignment.

File: src/cats/SlideAnimation.tsx [hunk 39]
```diff
@@ -0,0 +1,59 @@
...
+  child: JSX.Element;
+  direction: Direction;
+};
+
+function SlideAnimation({ child, direction }: SlideAnimationPropsType) {
+  const [slideChildren, setSlideChildren] = useState<Array<JSX.Element>>([]);
+
+  useEffect(() => {
+    setSlideChildren((prevChildren: Array<JSX.Element>) => {
+      const prevChild = prevChildren.slice(
```
Comment (ianobermiller) [inline src/cats/SlideAnimation.tsx:20]: consider, now that it is supported in newer browser:

```tsx
const prevChild = prevChildren.at(direction === Direction.Next ? -1 : 0);
```

Or for better support:

```tsx
const prevChild = prevChildren[direction === Direction.Next ? prevChidlren.length - 1 : 0];
```

Or the FB way:

```tsx
const prevChild = direction === Direction.Next ? last(prevChildren) : first(prevChildren);
```

File: src/cats/SlideAnimation.tsx [hunk 40]
```diff
@@ -0,0 +1,59 @@
...
+        ...(direction === Direction.Next ? [-1] : [0, 1])
+      );
+
+      return prevChild[0]?.key === child.key
+        ? prevChildren
+        : direction === Direction.Next
+        ? prevChild.concat([child])
```
Comment (ianobermiller) [inline src/cats/SlideAnimation.tsx:27]: if prevChild is not an array, this can just become `[prevChild. child]`

File: src/cats/SlideAnimation.tsx [hunk 41]
```diff
@@ -0,0 +1,59 @@
...
+
+  const onAnimationEnd = () => {
+    setSlideChildren([child]);
+  };
+
+  return (
+    <div className={styles.slideAnimation}>
+      {slideChildren.length === 2
+        ? [
+            slideChildren[0],
```
Comment (ianobermiller) [inline src/cats/SlideAnimation.tsx:40]: You might want to make sure this doesn't re-mount the child that is coming in when the animation completes. It's probably not worth the cost just to remove the wrapper.

File: src/cats/useAsync.ts [hunk 42]
```diff
@@ -0,0 +1,78 @@
...
+    case "loaded":
+      return { status: "loaded", data: action.data, error: null };
+    case "error":
+      return { status: "error", data: null, error: action.error };
+    case "abort":
+      return { status: "aborted", data: null, error: null };
+  }
+}
+
+export default function useAsync<D, E>(initialState: Status = "idle") {
```
Comment (ianobermiller) [inline src/cats/useAsync.ts:52]: nice usage of generics

File: src/cats/useAbortableFetch.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (jon-whiteroomsoftware) [reply src/cats/useAbortableFetch.js]: OK, now I understand what you meant. Thanks.

File: src/cats/BreedSelector.tsx (hunk 25 above)
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.tsx:6]: I guess if I should prefer to use interfaces over types, using the IFoo naming convention avoids having to come up with unique names for the interface and the function/component.

File: src/cats/BreedSelector.tsx (hunk 26 above)
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.tsx:11]: The performance benefit is only at compile time, it would make no difference at runtime, correct? I'm surprised it's substantial enough to warrant switching.

Do you still end up using types for discriminated unions, e.g:

```
interface Cash {
kind: "cash";
}

interface PayPal {
kind: "paypal";
email: string;
}

interface CreditCard {
kind: "creditcard";
cardNumber: string;
securityCode: string;
}

type PaymentMethod = Cash | PayPal | CreditCard;
```

File: src/cats/BreedSelector.tsx (hunk 28 above)
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.tsx:32]: Thanks, this is good to know.

File: src/cats/CatApp.tsx (hunk 34 above)
Comment (jon-whiteroomsoftware) [reply src/cats/CatApp.tsx:39]: good catch. I accidentally dropped the === undefined in the last diff. Will restore : )

File: src/cats/CatSlideshow.tsx (hunk 36 above)
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.tsx:110]: This was originally it's own effect, but I got some feedback from @chadwhiting that this component had too many side effects. My understanding was that for separation of concerns and performance reasons, it's best to use as many separate effects as you need.

I think the real win for this component was breaking the image prefetching out into it's own hook. It is a clean separation of functionality and has real reusability.

File: src/cats/BreedSelector.tsx (hunk 25 above)
Comment (ianobermiller) [reply src/cats/BreedSelector.tsx:6]: I think it doesn't matter too much the convention you use, as long as you stick to it. Wherever you go will probably have something in place already.

File: src/cats/BreedSelector.tsx (hunk 26 above)
Comment (ianobermiller) [reply src/cats/BreedSelector.tsx:11]: Yes, its only at typecheck time that you gain perf. We saw a 20% speedup in our codebase by converting a bunch of widely used core types to `interface`. And yes, types for unions is still necessary (part of the reason I prefer type everywhere).

File: src/cats/BreedSelector.tsx (hunk 28 above)
Comment (jon-whiteroomsoftware) [reply src/cats/BreedSelector.tsx:32]: so this useMemo is trying to avoid the work to recreate the options array on every render, not for stability.

File: src/cats/SlideAnimation.tsx (hunk 40 above)
Comment (jon-whiteroomsoftware) [reply src/cats/SlideAnimation.tsx:27]: I can do that but need to change a little more logic to make that work. The current code doesn't make it obvious that the when prevChildren is empty it ends up returning [child]. 

It would look something like this:

`const mostRecent = prevChildren.at(direction === "next" ? -1 : 0);

  return mostRecent === undefined
    ? [child]
    : mostRecent.key === child.key
    ? prevChildren
    : direction === "next"
    ? [mostRecent, child]
    : [child, mostRecent];`

File: src/cats/CatSlideshow.tsx (hunk 37 above)
Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.tsx:76]: I'm surprised you suggest passing callbacks into PaginatedFetch. Isn't the whole point to define state once and then write useEffects() to synchronize the propagation of state changes? At the simplest end, look at useLocalStateStorage. It defines its own state and returns the state to the caller. 

I do also see that employing useEffect is adding additional renders needlessly. The imperative callbacks just feel odd to me.

Comment (jon-whiteroomsoftware) [reply src/cats/CatSlideshow.tsx:76]: On further reflection, I might as well just pass the dispatch pointer from CatSlideshow into usePaginatedFetch. It's stable and it functions the same way you would if you passed it into a subcomponent.
